src/ast/codegen_llvm.cpp

```diff
@@ -2297,23 +2320,44 @@ void CodegenLLVM::createFormatStringCall(Call &call, int &id, CallArgs &call_arg
   for (size_t i=0; i<args.size(); i++)
   {
     Field &arg = args[i];
-    arg.offset = struct_layout->getElementOffset(i+1); // +1 for the id field
+    arg.offset = struct_layout->getElementOffset(i+1); // +1 for async action ID
   }
 
-  AllocaInst *fmt_args = b_.CreateAllocaBPF(fmt_struct, call_name + "_args");
-  b_.CREATE_MEMSET(fmt_args, b_.getInt8(0), struct_size, 1);
+  int asyncId = id + asyncactionint(async_action);
+  Value *fmt_args = b_.CreateGetFmtStrMap(ctx_, fmt_struct, call.loc);
+
+  Function *parent = b_.GetInsertBlock()->getParent();
+  BasicBlock *zero = BasicBlock::Create(module_->getContext(), "fmtstrzero", parent);
+  BasicBlock *notzero = BasicBlock::Create(module_->getContext(), "fmtstrnotzero", parent);
+
+  auto fmt_struct_ptr_ty = PointerType::get(fmt_struct, 0);
+  auto null_ptr = ConstantExpr::getCast(Instruction::IntToPtr, b_.getInt64(0), fmt_struct_ptr_ty);
+  b_.CreateCondBr(b_.CreateICmpNE(fmt_args, null_ptr, "fmtstrcond"), notzero, zero);
+
+  b_.SetInsertPoint(notzero);
+
+  auto zeroed_area_ptr = b_.getInt64(reinterpret_cast<uintptr_t>(bpftrace_.zero_buffer_->data()));
+
+  b_.CreateProbeRead(ctx_, fmt_args, struct_size,
+                     ConstantExpr::getCast(Instruction::IntToPtr, zeroed_area_ptr, fmt_struct_ptr_ty), call.loc);
 
   Value *id_offset = b_.CreateGEP(fmt_args, {b_.getInt32(0), b_.getInt32(0)});
-  b_.CreateStore(b_.getInt64(id + asyncactionint(async_action)), id_offset);
+  b_.CreateStore(b_.getInt64(asyncId), id_offset);
   for (size_t i=1; i<call.vargs->size(); i++)
   {
     Expression &arg = *call.vargs->at(i);
     expr_deleter_ = nullptr;
+    expr_points_to_map_value = false;
     arg.accept(*this);
     Value *offset = b_.CreateGEP(fmt_args, {b_.getInt32(0), b_.getInt32(i)});
+    // TODO: consider checking whether expr_ is a nullptr to map value
     if (arg.type.IsAggregate())
     {
-      b_.CREATE_MEMCPY(offset, expr_, arg.type.size, 1);
+      if (expr_points_to_map_value) {
+        b_.CreateProbeRead(ctx_, offset, arg.type.size, expr_, call.loc);
+      } else {
+        b_.CREATE_MEMCPY(offset, expr_, arg.type.size, 1);
+      }
```

Comment:
Shouldn't these arms be flipped? If it's in a map we can do a regular memcpy right?

Reply:
I was surprised, but MEMCPY from str()'s map to printf()'s map failed as invalid. something to do with `inv`. I can grab the message if you'd like to see? but this way round was definitely necessary to fix it.

Reply:
Looking at this again, this looks right. Not sure what I was thinking before